File: webapp/routers/shares.py
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel
from typing import List, Optional
from services.config_parser import SmbConfParser
from services.docker_executor import DockerExecutor
from services.audit_service import AuditService
from config import settings
import os
import logging

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/shares", tags=["shares"])

class ShareCreate(BaseModel):
    name: str
    path: str
    comment: str = ""
    browseable: bool = True
    guest_ok: bool = False
    abse: bool = False
    create_mask: str = "0755"
    directory_mask: str = "0755"

class ShareUpdate(BaseModel):
    comment: Optional[str] = None
    browseable: Optional[bool] = None
    guest_ok: Optional[bool] = None
    read_only: Optional[bool] = None
    abse: Optional[bool] = None
    create_mask: Optional[str] = None
    directory_mask: Optional[str] = None

class PermissionUpdate(BaseModel):
    users: List[str] = []
    permission_type: str

class GlobalSettings(BaseModel):
    abse: bool

def get_config():
    return SmbConfParser(os.path.join(settings.samba_config_path, 'smb.conf'))

def get_executor():
    return DockerExecutor(settings.samba_container)

def get_actor(request: Request) -> str:
    return getattr(request.state, 'username', 'unknown')

# ⚠️ Static routes BEFORE /{share_name}

@router.get("/global")
async def get_global_settings():
    try:
        return get_config().get_global()
    except Exception as e:
        raise HTTPException(500, detail=str(e))

@router.patch("/global")
async def update_global_settings(body: GlobalSettings, request: Request):
    try:
        cfg = get_config()
        cfg.set_global_abse(body.abse)
        get_executor().reload_samba()
        AuditService.log(
            action="UPDATE_GLOBAL", actor=get_actor(request),
            resource_type="SHARE", resource_name="global",
            status="success", details={"abse": body.abse}
        )
        return {"message": f"Global ABSE set to: {body.abse}"}
    except Exception as e:
        raise HTTPException(500, detail=str(e))

@router.get("")
async def list_shares():
    try:
        return {"shares": get_config().get_all_shares()}
    except Exception as e:
        raise HTTPException(500, detail=str(e))

@router.post("")
async def create_share(share: ShareCreate, request: Request):
    try:
        ex = get_executor()
        ex.execute(f"mkdir -p {share.path} && chmod 777 {share.path}")
        cfg = get_config()
        if not cfg.create_share(share.name, share.path, share.comment):
            raise HTTPException(400, "Share already exists")
        cfg.update_share(share.name, {
            "browseable": share.browseable,
            "guest_ok": share.guest_ok,
            "abse": share.abse,
            "create mask": share.create_mask,
            "directory mask": share.directory_mask,
        })
        ex.reload_samba()
        AuditService.log(
            action="CREATE", actor=get_actor(request),
            resource_type="SHARE", resource_name=share.name,
            status="success", details={"path": share.path}
        )
        return {"message": f"Share '{share.name}' created"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(500, detail=str(e))

@router.patch("/{share_name}/abse")
async def toggle_share_abse(share_name: str, enabled: bool, request: Request):
    try:
        cfg = get_config()
        if not cfg.get_share(share_name):
            raise HTTPException(404, f"Share '{share_name}' not found")
        cfg.set_share_abse(share_name, enabled)
        get_executor().reload_samba()
        AuditService.log(
            action="UPDATE", actor=get_actor(request),
            resource_type="SHARE", resource_name=share_name,
            status="success", details={"abse": enabled}
        )
        return {"message": f"ABSE {'enabled' if enabled else 'disabled'} for '{share_name}'"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(500, detail=str(e))

@router.post("/{share_name}/permissions")
async def update_permissions(share_name: str, perm: PermissionUpdate, request: Request):
    try:
        cfg = get_config()
        if not cfg.get_share(share_name):
            raise HTTPException(404, "Share not found")
        perm_map = {
            'valid_users':   cfg.set_valid_users,
            'write_list':    cfg.set_write_list,
            'read_list':     cfg.set_read_list,
            'admin_users':   cfg.set_admin_users,
            'invalid_users': cfg.set_invalid_users,
        }
        if perm.permission_type not in perm_map:
            raise HTTPException(400, f"Invalid permission_type: {perm.permission_type}")
        perm_map[perm.permission_type](share_name, perm.users)
        get_executor().reload_samba()
        AuditService.log(
            action="PERMISSION_CHANGE", actor=get_actor(request),
            resource_type="SHARE", resource_name=share_name,
            status="success", details={"type": perm.permission_type, "users": perm.users}
        )
        return {"message": "Permissions updated"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(500, detail=str(e))

@router.get("/{share_name}")
async def get_share(share_name: str):
    try:
        share = get_config().get_share(share_name)
        if not share:
            raise HTTPException(404, "Share not found")
        return share
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(500, detail=str(e))

@router.patch("/{share_name}")
async def update_share(share_name: str, updates: ShareUpdate, request: Request):
    try:
        cfg = get_config()
        if not cfg.get_share(share_name):
            raise HTTPException(404, "Share not found")
        update_dict = {k: v for k, v in updates.dict().items() if v is not None}
        cfg.update_share(share_name, update_dict)
        get_executor().reload_samba()
        AuditService.log(
            action="UPDATE", actor=get_actor(request),
            resource_type="SHARE", resource_name=share_name,
            status="success", details=update_dict
        )
        return {"message": f"Share '{share_name}' updated"}
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(500, detail=str(e))

@router.delete("/{share_name}")
async def delete_share(share_name: str, request: Request):
    try:
        get_config().delete_share(share_name)
        get_executor().reload_samba()
        AuditService.log(
            action="DELETE", actor=get_actor(request),
            resource_type="SHARE", resource_name=share_name,
            status="success"
        )
        return {"message": f"Share '{share_name}' deleted"}
    except Exception as e:
        raise HTTPException(500, detail=str(e))

File: backend/internal/httpapi/shares.go
package httpapi

import (
	"regexp"

	"smb-server/backend/internal/samba"

	"github.com/gofiber/fiber/v2"
)

// ValidSharePath allows absolute paths with safe characters only — no shell metacharacters.
var ValidSharePath = regexp.MustCompile(`^/[a-zA-Z0-9/_-]+$`)

type sharesHandlers struct {
	exec       samba.Executor
	configPath string
}

type shareCreateRequest struct {
	Name          string `json:"name"`
	Path          string `json:"path"`
	Comment       string `json:"comment"`
	Browseable    bool   `json:"browseable"`
	GuestOK       bool   `json:"guest_ok"`
	ABSE          bool   `json:"abse"`
	CreateMask    string `json:"create_mask"`
	DirectoryMask string `json:"directory_mask"`
}

type shareUpdateRequest struct {
	Comment       *string `json:"comment"`
	Browseable    *bool   `json:"browseable"`
	GuestOK       *bool   `json:"guest_ok"`
	ReadOnly      *bool   `json:"read_only"`
	ABSE          *bool   `json:"abse"`
	CreateMask    *string `json:"create_mask"`
	DirectoryMask *string `json:"directory_mask"`
}

type permissionUpdateRequest struct {
	Users          []string `json:"users"`
	PermissionType string   `json:"permission_type"`
}

type abseRequest struct {
	Enabled bool `json:"enabled"`
}

func registerSharesRoutes(g fiber.Router, exec samba.Executor, configPath string) {
	h := &sharesHandlers{exec: exec, configPath: configPath}
	// Static routes before /:name
	g.Get("/global", h.getGlobal)
	g.Patch("/global", h.updateGlobal)
	g.Get("", h.list)
	g.Post("", h.create)
	g.Patch("/:name/abse", h.toggleABSE)
	g.Post("/:name/permissions", h.updatePermissions)
	g.Get("/:name", h.get)
	g.Patch("/:name", h.update)
	g.Delete("/:name", h.delete)
}

func (h *sharesHandlers) parser() *samba.SmbConfParser {
	return samba.NewSmbConfParser(h.configPath+"/shares.conf", h.configPath+"/smb.conf")
}

func (h *sharesHandlers) getGlobal(c *fiber.Ctx) error {
	return c.JSON(h.parser().GetGlobal())
}

func (h *sharesHandlers) updateGlobal(c *fiber.Ctx) error {
	// Global config is template-managed; no-op but still reload
	h.exec.ReloadSamba()
	return c.JSON(fiber.Map{"message": "Global config is managed by template"})
}

func (h *sharesHandlers) list(c *fiber.Ctx) error {
	shares, err := h.parser().GetAllShares()
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"detail": err.Error()})
	}
	return c.JSON(fiber.Map{"shares": shares})
}

func (h *sharesHandlers) create(c *fiber.Ctx) error {
	var req shareCreateRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(400).JSON(fiber.Map{"detail": "invalid request body"})
	}
	if !ValidSharePath.MatchString(req.Path) {
		return c.Status(400).JSON(fiber.Map{"detail": "invalid path: must be absolute and contain only [a-zA-Z0-9/_-]"})
	}
	// mkdir + chmod in samba container (path is validated above — no injection risk)
	h.exec.Execute("mkdir -p " + req.Path + " && chmod 770 " + req.Path)

	p := h.parser()
	if !p.CreateShare(req.Name, req.Path, req.Comment) {
		return c.Status(400).JSON(fiber.Map{"detail": "Share already exists"})
	}
	updates := map[string]interface{}{
		"browseable":     req.Browseable,
		"guest_ok":       req.GuestOK,
		"abse":           req.ABSE,
		"create mask":    req.CreateMask,
		"directory mask": req.DirectoryMask,
	}
	p.UpdateShare(req.Name, updates)
	h.exec.ReloadSamba()
	return c.Status(201).JSON(fiber.Map{"message": "Share '" + req.Name + "' created"})
}

func (h *sharesHandlers) get(c *fiber.Ctx) error {
	name := c.Params("name")
	share, err := h.parser().GetShare(name)
	if err != nil || share == nil {
		return c.Status(404).JSON(fiber.Map{"detail": "Share not found"})
	}
	return c.JSON(share)
}

func (h *sharesHandlers) update(c *fiber.Ctx) error {
	name := c.Params("name")
	var req shareUpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(400).JSON(fiber.Map{"detail": "invalid request body"})
	}
	p := h.parser()
	if ok, _ := p.ShareExists(name); !ok {
		return c.Status(404).JSON(fiber.Map{"detail": "Share not found"})
	}
	updates := map[string]interface{}{}
	if req.Comment != nil {
		updates["comment"] = *req.Comment
	}
	if req.Browseable != nil {
		updates["browseable"] = *req.Browseable
	}
	if req.GuestOK != nil {
		updates["guest_ok"] = *req.GuestOK
	}
	if req.ReadOnly != nil {
		updates["read_only"] = *req.ReadOnly
	}
	if req.ABSE != nil {
		updates["abse"] = *req.ABSE
	}
	if req.CreateMask != nil {
		updates["create mask"] = *req.CreateMask
	}
	if req.DirectoryMask != nil {
		updates["directory mask"] = *req.DirectoryMask
	}
	p.UpdateShare(name, updates)
	h.exec.ReloadSamba()
	return c.JSON(fiber.Map{"message": "Share '" + name + "' updated"})
}

func (h *sharesHandlers) delete(c *fiber.Ctx) error {
	name := c.Params("name")
	h.parser().DeleteShare(name)
	h.exec.ReloadSamba()
	return c.JSON(fiber.Map{"message": "Share '" + name + "' deleted"})
}

func (h *sharesHandlers) toggleABSE(c *fiber.Ctx) error {
	name := c.Params("name")
	enabled := c.QueryBool("enabled")
	p := h.parser()
	if ok, _ := p.ShareExists(name); !ok {
		return c.Status(404).JSON(fiber.Map{"detail": "Share not found"})
	}
	p.SetShareABSE(name, enabled)
	h.exec.ReloadSamba()
	return c.JSON(fiber.Map{"message": "ABSE updated"})
}

func (h *sharesHandlers) updatePermissions(c *fiber.Ctx) error {
	name := c.Params("name")
	var req permissionUpdateRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(400).JSON(fiber.Map{"detail": "invalid request body"})
	}
	p := h.parser()
	if ok, _ := p.ShareExists(name); !ok {
		return c.Status(404).JSON(fiber.Map{"detail": "Share not found"})
	}
	validTypes := map[string]func(string, []string) bool{
		"valid_users":   p.SetValidUsers,
		"write_list":    p.SetWriteList,
		"read_list":     p.SetReadList,
		"admin_users":   p.SetAdminUsers,
		"invalid_users": p.SetInvalidUsers,
	}
	setter, ok := validTypes[req.PermissionType]
	if !ok {
		return c.Status(400).JSON(fiber.Map{"detail": "Invalid permission_type: " + req.PermissionType})
	}
	setter(name, req.Users)
	h.exec.ReloadSamba()
	return c.JSON(fiber.Map{"message": "Permissions updated"})
}
